### PR TITLE
dev/core#192 - Search builder fails for != smart group filter

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Overview
----------------------------------------
Fix smart group in search builder with `!=` operator.

Before
----------------------------------------
Smart Group search on `Search Builder` result incorrect rows with `!=` operator

image

After
----------------------------------------
Correct results are displayed.

Comments
----------------------------------------
Added unit test.

----------------------------------------

Gitlab - https://lab.civicrm.org/dev/core/issues/192

### CODE REVIEW

**Comment by monishdeb** [comment]
@jitendrapurohit I encountered an issue, with this patch. Let's say Contact A and B is in regular group RG AND contact B and C is smart group SG. Then when I do a search with 'Group != SG' then it returns all the 4 contacts. Shouldn't we expect only Contact A in the result list?

On debugging I found that at this [query](https://github.com/civicrm/civicrm-core/pull/12344/files#diff-e54381bfdf51e31cab376c71ca0d66ffR3120) it excludes the smart group ID and eventually didn't build any where clause for the chosen smart group. 

But the intent of this patch is to make = and != operator works like IN/NOT IN respectively for smart group, but the use-case described above bothers me to as it still includes the contact in the result list, which is in SG

**Comment by monishdeb** [comment]
Considering A, B, C and D are 4 contacts, where Contact A and B is in regular group RG AND contact B and C is smart group SG, C and D are in second SG2. Here's my test cases after applying the patches:
1. **Group = IS NULL** : Return all contacts which is not added in any group - correct
2. **Group = SG**: Return B and C - correct
3. **Group != SG**: Return A,B,C,D - is it expected? or as per the fix if a contact is ONLY present in SG then it should be excluded from the list BUT if the contact is present is more then 1 smart grouo that it is not? The underlying where clause is:
```
// here 6 is the group ID of SG
`civicrm_group_contact-6`.group_id != 6 OR  ( `civicrm_group_contact_cache_6`.group_id NOT IN ("6") )
```
4.  **Group NOT IN (SG)** : Return A,B,C,D produce same result as of case3. But there is difference in where clause fragment:
````civicrm_group_contact-6`.group_id NOT IN ( 6 ) OR  ( `civicrm_group_contact_cache_6`.group_id NOT IN ("6") )
```
Here we should exclude the OR operator which is expensive in terms of performance and both the sub-clause is same.
5. **Group = IS NOT NULL** : Return only those contacts which are added in Group.

@jitendrapurohit so overall : 
1. Can you fix the where clause as reported in 4rth case.
2. Test failure seems to be related, can you fix those 2?
3. Can you extend your added UT to check with Contacts other then substring check for whereClause?

**Comment by jitendrapurohit** [comment]
Updated the PR and handled the NOT operator clauses in the same way as we do for `civicrm_group_contact` table. Tested all the above scenarios and looks fine on my local. Also updated the unit test. Can you verify if it works fine for you as well. Thanks @monishdeb

**Comment by monishdeb** [comment]
@jitendrapurohit it seems that the 1 test failure is related to this patch. Can you fix that?

**Comment by jitendrapurohit** [comment]
weird, the failing test is the one added in this PR and it works fine on my local. Anyway, I've pushed a commit which should fix the same on jenkin 👍

**Comment by jitendrapurohit** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
(re-running tests as 22 days ago)

**Comment by eileenmcnaughton** [comment]
@monishdeb looks like this needs a final ok from you

**Comment by monishdeb** [comment]
@eileenmcnaughton I am not sure in this case, as for smart group, it adds a subquery when  != or NOT IN operator is used, which will cause a performance issue. As per the patch when I do search on Group != Smart Group (id = 10) it results into:
```
...
WHERE (  (  ( `civicrm_group_contact-10`.group_id != 10 AND `civicrm_group_contact_cache_10`.group_id IS NULL OR  ( `civicrm_group_contact_cache_10`.contact_id NOT IN (SELECT contact_id FROM civicrm_group_contact_cache cgcc WHERE cgcc.group_id IN ( 10 ) ) )  )  )  ) 
```
 The use of OR operator and subquery will cause a performance hit on large site. As I recall, I had a discussion with you earlier to have separate LEFT JOINS for smart group and regular groups and in this way we can resolved the performance issue for smart and regular group. Is it time to bring this change in core?

**Comment by eileenmcnaughton** [comment]
Hmm - yeah - we should get some performance testing for sure! @seamuslee001 has a good site for testing group search performance on

**Comment by jackrabbithanna** [comment]
I've tried this out on a client site which we had found the accuracy problem while testing a 4.6 -> 5.6 upgrade. Upgraded to 5.7 now and applied this patch, and it actually get's the right results and doesn't WSOD.  Not exactly sure out this site compares to Seamus's site...but we got not quite 100K contacts, several hundred smart groups, smart groups based on activities, with millions of activities....and logging enabled...At the very least, this is better than it was with 5.6 and no patch!

**Comment by eileenmcnaughton** [comment]
@seamuslee001 Any chance you can give it a quick spin to double check? 

Looking at the query I think the OR might not be the problem OR that gave all those WSODs in 4.6 -  in that it is in the WHERE not the join and it's 2 fields on the same table.

**Comment by jackrabbithanna** [comment]
I was testing with 4 criteria, using 3 there where contact = group, and one where contact != group..

All groups were smart groups. It was a little slow, takes about 220s to return the search results....so by better than not working, yeah, but pretty slow...I'll take really slow over nothing at all any day though.

All the groups use criteria looking for if a contact has an activity of a certain type, between certain activity_date_times , with certain activity status...

MySQL 5.5 AWS EB instance, no memcache, no varnish, no redis...

File: CRM/Contact/BAO/Query.php [hunk 1]
```diff
@@ -3048,12 +3048,33 @@ public function group($values) {
     }
 
     //CRM-19589: contact(s) removed from a Smart Group, resides in civicrm_group_contact table
-    if (count($smartGroupIDs)) {
-      $groupClause[] = " ( " . $this->addGroupContactCache($smartGroupIDs, NULL, "contact_a", $op) . " ) ";
+    $groupContactCacheClause = '';
+    if (count($smartGroupIDs) || empty($value)) {
+      $isNullOp = (strpos($op, 'NULL') !== FALSE);
+      $gccTableAlias = "`civicrm_group_contact_cache_";
+      $gccTableAlias .= ($isNullOp) ? "a`" : implode(',', $smartGroupIDs) . "`";
```
[the remark this answers is not in the document]
Comment (jitendrapurohit) [reply CRM/Contact/BAO/Query.php]: Updated the changes and also tested it on Advanced search and Search builder screen. Thanks for pushing this further 👍

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit the test works :-)

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by jitendrapurohit** [comment]
jenkins, retest this please

File: CRM/Contact/BAO/Query.php [hunk 2]
```diff
@@ -3041,12 +3041,32 @@ public function group($values) {
     }
 
     //CRM-19589: contact(s) removed from a Smart Group, resides in civicrm_group_contact table
-    if (count($smartGroupIDs)) {
-      $groupClause[] = " ( " . $this->addGroupContactCache($smartGroupIDs, NULL, "contact_a", $op) . " ) ";
+    $groupContactCacheClause = '';
+    if (count($smartGroupIDs) || empty($value)) {
+      $isNullOp = (strpos($op, 'NULL') !== FALSE);
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: Why are we not just doing
```
 ($op == 'IS NULL') 
```

- we don't support people playing fast & loose with the $op var do we?

Comment (jitendrapurohit) [reply CRM/Contact/BAO/Query.php]: Hmm, I see this variable is not required in this function after the latest changes. Pushed the commit to remove this line completely. Thanks for identifying.

File: CRM/Contact/BAO/Query.php [hunk 3]
```diff
@@ -3041,12 +3041,31 @@ public function group($values) {
...
-    if (count($smartGroupIDs)) {
-      $groupClause[] = " ( " . $this->addGroupContactCache($smartGroupIDs, NULL, "contact_a", $op) . " ) ";
+    $groupContactCacheClause = '';
+    if (count($smartGroupIDs) || empty($value)) {
+      $gccTableAlias = "civicrm_group_contact_cache";
+      $groupContactCacheClause = $this->addGroupContactCache($smartGroupIDs, $gccTableAlias, "contact_a", $op);
+      if (!empty($groupContactCacheClause)) {
+        if ($isNotOp) {
+          $groupIds = implode(',', (array) $smartGroupIDs);
+          $gcTable = "`civicrm_group_contact-{$groupIds}`";
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: @jitendrapurohit don't we still have appended input here?

Comment (jitendrapurohit) [reply CRM/Contact/BAO/Query.php]: There are still some places in civi where id is appended to the group table and was little sceptic about removing it here. But it looks like they are not dependent on changes made in this file so i've just removed it now. will wait for the test build to run clean this time too.

**Comment by jitendrapurohit** [comment]
test this please

**Comment by jitendrapurohit** [comment]
With the last change in removing id from `civicrm_group_contact-` - the test that fails is -

    $group = civicrm_api3('Contact', 'get', [
        'sequential' => 1,
        'return' => ["group"],
        'group' => array(group_title => 1),
      ]);
      
i.e, searching for contact with group title should retrieve all the group for the contacts which was done with the additional alias added to the query. 

The latest update to this PR adds uniqid to the table name in the from clause which fixes it and retrieves correct number of groups in the result.

**Comment by jitendrapurohit** [comment]
jenkins, retest this please

**Comment by eileenmcnaughton** [comment]
woot tests are passing !!

@jackrabbithanna do you want to give this a last spin & then we can merge if all good

**Comment by jackrabbithanna** [comment]
Sorry, I just noticed you pinged me on this.....I'll test this over the weekend and let you know....

**Comment by eileenmcnaughton** [comment]
@jackrabbithanna thanks - I'm going to merge this now since it has been tested by you more than once now & also has unit tests. It will be another 6 weeks before it goes out live - during which time you'll be testing further - so please let us know any more updates from your testing

**Comment by eileenmcnaughton** [comment]
This turned out tricksy - we also got https://github.com/civicrm/civicrm-core/pull/13662 - the underlying issue is poor test coverage of a fairly tricky bit of functionality. This change DID improve that & @seamuslee001 has promised to add follow up tests but hoping you can check the new PR @jitendrapurohit

**Pull request merged**